File: extensions/poncho-wb-parser/src/storage/mock.ts
// src/storage/mock.ts — synthetic captures for testing Decode + the collect loop without the
// browser or live WB. Port of MockIntercepts() in pkg/wbscraper/mock.go.
// Bodies are JS objects (already-parsed — the in-extension interceptor JSON.parse's before forwarding).
//
// The search capture sits on page=2 with dest=8038 so position math is non-trivial
// ((2-1)*100+idx+1), and mixes an organic listing (panelPromoId null) with an ad (panelPromoId 99).

import type { Intercept } from '../db/types';

export function mockIntercepts(): Intercept[] {
  return [
    {
      kind: 'search',
      url: 'https://www.wildberries.ru/catalog/0/search.aspx?search=кроссовки&page=2&dest=8038',
      query_id: 7,
      status: 200,
      body: {
        metadata: { name: 'кроссовки' },
        products: [
          {
            id: 111,
            brand: 'Nike',
            supplierId: 900,
            panelPromoId: null,
            rating: 4.5,
            feedbacks: 10,
            sizes: [{ price: { basic: 100000, product: 89900 } }],
          },
          {
            id: 222,
            brand: 'Adidas',
            supplierId: 901,
            panelPromoId: 99,
            rating: 4.0,
            feedbacks: 5,
            sizes: [{ price: { basic: 50000, product: 45000 } }],
          },
        ],
      },
    },
    {
      kind: 'card_detail',
      url: 'https://www.wildberries.ru/catalog/111/detail.aspx',
      query_id: 7,
      status: 200,
      body: {
        products: [
          {
            id: 111,
            brand: 'Nike',
            supplier: 'ООО Рога',
            supplierId: 900,
            rating: 4.5,
            feedbacks: 10,
            pics: ['a.jpg', 'b.jpg'],
            colors: [{ name: 'черный' }],
            subjectId: 81,
            panelPromoId: null,
            totalQuantity: 250,
            promotions: [{ name: 'Скидка' }],
            sizes: [
              {
                name: '42',
                price: { basic: 100000, product: 89900 },
                stocks: [{ wh: 507, qty: 10, time1: 1720000000, time2: 1720003600 }],
              },
            ],
          },
        ],
      },
    },
    {
      // wbbasket.ru static CDN card.json — flat content object (description + characteristics).
      // nm_id is in the body (authoritative); the URL path also encodes it as a fallback.
      kind: 'card_content',
      url: 'https://basket-10.wbbasket.ru/vol1/part111/111/info/ru/card.json',
      query_id: 7,
      status: 200,
      body: {
        imt_id: 999,
        nm_id: 111,
        imt_name: 'Кроссовки беговые мужские',
        slug: 'krossovki-begovye-muzhskie',
        vendor_code: '22123456',
        subj_name: 'Кроссовки',
        subj_root_name: 'Обувь',
        description: 'Беговые кроссовки',
        markdown_description: 'Беговые кроссовки **мужские**',
        need_kiz: true,
        create_date: '2022-03-01T00:00:00Z',
        update_date: '2026-06-01T00:00:00Z',
        options: [
          { name: 'Состав', value: 'текстиль 100%', charc_type: 1 },
          { name: 'Цвет', value: 'черный', charc_type: 1, is_variable: true, variable_values: ['черный', 'белый'] },
          { name: 'Вид спорта', value: 'бег', charc_type: 1 },
        ],
        compositions: [{ name: 'текстиль 100%' }],
        sizes_table: {
          details_props: ['RU', 'Длина стельки, см'],
          values: [
            { tech_size: '39', chrt_id: 134002594, details: ['39', '25'] },
            { tech_size: '40', chrt_id: 134002595, details: ['40', ''] }, // sparse: empty cell skipped
          ],
        },
        colors: [111, 222333],
        nm_colors_names: 'черный',
        contents: 'Кроссовки 1 шт',
        kinds: ['Мужчины'],
        has_seller_recommendations: true,
        user_flags: 1,
        selling: { brand_name: 'Nike', brand_hash: 'ABC123', supplier_id: 900 },
        media: { photo_count: 6, has_video: true },
        data: { subject_id: 81, subject_root_id: 4, chrt_ids: [134002594, 134002595] },
        grouped_options: [
          { group_name: 'Основная информация', options: [{ name: 'Состав', value: 'текстиль 100%', charc_type: 1 }, { name: 'Цвет', value: 'черный', charc_type: 1 }] },
          { group_name: 'Дополнительная информация', options: [{ name: 'Вид спорта', value: 'бег', charc_type: 1 }] },
        ],
      },
    },
  ];
}

File: pkg/wbscraper/mock.go
package wbscraper

import (
	"context"
	"encoding/json"
	"sync"
)

// DiscardWriter implements Writer with no-op persistence. Used in --mock mode to
// guarantee zero DB interaction (Writer creation goes inside the non-mock branch,
// per the DiscardWriter pattern — see CLAUDE.md "V2 Downloader Architecture").
//
// UpsertQuery is faithful, not vacuous: it assigns synthetic ids and records the
// query text so identical text maps to one id across calls (mirroring the real DB's
// UNIQUE constraint). This lets --mock mode exercise the server's query-stamping
// logic end-to-end — targets still get stable QueryIDs, captures still carry them —
// without touching storage.
type DiscardWriter struct {
	mu sync.Mutex

	// synthetic query ids (idempotent by query text, like a DB UNIQUE).
	nextQueryID int64
	queries     map[string]int64

	// per-table "saved" counters, for test assertions (parity with pkg/supplies).
	savedPositions  int
	savedVitrineAds int
	savedCards      int
	savedPrices     int
	savedDetails    int
	savedStocks     int
}

// NewDiscardWriter creates a no-op writer for mock mode.
func NewDiscardWriter() *DiscardWriter {
	return &DiscardWriter{queries: make(map[string]int64)}
}

// UpsertQuery returns the existing synthetic id for q.Query, or assigns the next
// one. Idempotent: the same text always yields the same id, as a real
// search_queries UNIQUE would. Ids start at 1, so NoQuery (0) stays reserved.
func (w *DiscardWriter) UpsertQuery(_ context.Context, q SearchQuery) (int64, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	if id, ok := w.queries[q.Query]; ok {
		return id, nil
	}
	w.nextQueryID++
	w.queries[q.Query] = w.nextQueryID
	return w.nextQueryID, nil
}

// SaveStorefrontPositions counts rows but never writes to any database.
func (w *DiscardWriter) SaveStorefrontPositions(_ context.Context, rows []SearchPosition) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedPositions += len(rows)
	return len(rows), nil
}

// SaveVitrineAds counts rows but never writes.
func (w *DiscardWriter) SaveVitrineAds(_ context.Context, rows []VitrineAd) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedVitrineAds += len(rows)
	return len(rows), nil
}

// SaveCompetitorCards counts rows but never writes.
func (w *DiscardWriter) SaveCompetitorCards(_ context.Context, rows []CompetitorCard) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedCards += len(rows)
	return len(rows), nil
}

// SaveCompetitorCardPrices counts rows but never writes.
func (w *DiscardWriter) SaveCompetitorCardPrices(_ context.Context, rows []CompetitorCardPrice) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedPrices += len(rows)
	return len(rows), nil
}

// SaveCompetitorCardDetails counts rows but never writes.
func (w *DiscardWriter) SaveCompetitorCardDetails(_ context.Context, rows []CompetitorCardDetail) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedDetails += len(rows)
	return len(rows), nil
}

// SaveCompetitorCardStocks counts rows but never writes.
func (w *DiscardWriter) SaveCompetitorCardStocks(_ context.Context, rows []CompetitorCardStock) (int, error) {
	w.mu.Lock()
	defer w.mu.Unlock()
	w.savedStocks += len(rows)
	return len(rows), nil
}

// SavedQueries returns the number of distinct query texts upserted.
func (w *DiscardWriter) SavedQueries() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return len(w.queries)
}

// SavedStorefrontPositions returns the count of position rows "saved" (counted).
func (w *DiscardWriter) SavedStorefrontPositions() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedPositions
}

// SavedVitrineAds returns the count of ad rows "saved" (counted).
func (w *DiscardWriter) SavedVitrineAds() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedVitrineAds
}

// SavedCompetitorCards returns the count of card rows "saved" (counted).
func (w *DiscardWriter) SavedCompetitorCards() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedCards
}

// SavedCompetitorCardPrices returns the count of price rows "saved" (counted).
func (w *DiscardWriter) SavedCompetitorCardPrices() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedPrices
}

// SavedCompetitorCardDetails returns the count of detail rows "saved" (counted).
func (w *DiscardWriter) SavedCompetitorCardDetails() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedDetails
}

// SavedCompetitorCardStocks returns the count of stock rows "saved" (counted).
func (w *DiscardWriter) SavedCompetitorCardStocks() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.savedStocks
}

// Compile-time assertion that DiscardWriter satisfies Writer (Stage 3 storage
// adapters carry the same assertion).
var _ Writer = (*DiscardWriter)(nil)

// MockIntercepts returns synthetic captures (one search + one card_detail) for
// unit-testing Decode and the server loop without the browser extension or fixture
// files. Bodies match the WB response shapes documented in decode.go. The search
// capture sits on page=2 with dest=8038 so position math is non-trivial
// ((2-1)*100+idx+1), and mixes an organic listing (panelPromoId null) with an ad.
func MockIntercepts() []Intercept {
	return []Intercept{
		{
			Kind:    "search",
			URL:     "https://www.wildberries.ru/catalog/0/search.aspx?search=кроссовки&page=2&dest=8038",
			QueryID: 7,
			Status:  200,
			Body: json.RawMessage(`{
  "metadata": {"name": "кроссовки"},
  "products": [
    {"id": 111, "brand": "Nike", "supplierId": 900, "panelPromoId": null, "rating": 4.5, "feedbacks": 10,
     "sizes": [{"price": {"basic": 100000, "product": 89900}}]},
    {"id": 222, "brand": "Adidas", "supplierId": 901, "panelPromoId": 99, "rating": 4.0, "feedbacks": 5,
     "sizes": [{"price": {"basic": 50000, "product": 45000}}]}
  ]
}`),
		},
		{
			Kind:    "card_detail",
			URL:     "https://www.wildberries.ru/catalog/111/detail.aspx",
			QueryID: 7,
			Status:  200,
			Body: json.RawMessage(`{
  "products": [
    {"id": 111, "brand": "Nike", "supplier": "ООО Рога", "supplierId": 900, "rating": 4.5, "feedbacks": 10,
     "pics": ["a.jpg","b.jpg"], "colors": [{"name":"черный"}], "subjectId": 81,
     "panelPromoId": null, "totalQuantity": 250, "promotions": [{"name":"Скидка"}],
     "sizes": [
       {"name": "42", "price": {"basic": 100000, "product": 89900},
        "stocks": [{"wh": 507, "qty": 10, "time1": 1720000000, "time2": 1720003600}]}
     ]}
  ]
}`),
		},
	}
}
